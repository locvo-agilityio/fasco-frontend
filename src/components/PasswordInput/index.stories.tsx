import type { Meta, StoryObj } from '@storybook/react';

// Components
import PasswordInput from '.';

const meta = {
  title: 'Components/PasswordInput',
  component: PasswordInput,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof PasswordInput>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    placeholder: 'John Carter',
  },
};

export const HasError: Story = {
  args: {
    isInvalid: true,
    errorMessage: 'Error message',
    placeholder: 'John Carter',
    defaultValue: 'Error Data',
  },
};
