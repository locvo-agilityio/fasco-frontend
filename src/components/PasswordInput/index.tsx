'use client';

import { ComponentProps, memo, useState } from 'react';
import { Eye, EyeOff } from 'lucide-react';

// Component
import { Button, Input } from '@/components/common';

// Utils
import { cn } from '@/utils';

const PasswordInput = ({
  isInvalid,
  ...rest
}: ComponentProps<typeof Input>) => {
  const [isVisible, setIsVisible] = useState(false);

  const toggleVisibility = () => setIsVisible((prev) => !prev);

  return (
    <Input
      endContent={
        <Button
          title="toggle password visibility"
          color="primary"
          size="md"
          variant="ghost"
          aria-label="toggle password visibility"
          className="w-fit focus:outline-none border-none data-[hover=true]:!bg-transparent"
          type="button"
          onPress={toggleVisibility}
        >
          {isVisible ? (
            <Eye
              size={20}
              className={cn(
                'pointer-events-none',
                isInvalid ? 'text-danger-300' : 'text-default-400',
              )}
            />
          ) : (
            <EyeOff
              size={20}
              className={cn(
                'pointer-events-none',
                isInvalid ? 'text-danger-300' : 'text-default-400',
              )}
            />
          )}
        </Button>
      }
      type={isVisible ? 'text' : 'password'}
      isInvalid={isInvalid}
      {...rest}
    />
  );
};

export default memo(PasswordInput);
