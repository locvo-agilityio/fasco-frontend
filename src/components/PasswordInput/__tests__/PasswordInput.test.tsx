import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Components
import PasswordInput from '..';

describe('PasswordInput component', () => {
  const mockOnChange = jest.fn();
  const props = {
    label: 'Email',
    name: 'email',
    placeholder: 'Email',
    onChange: mockOnChange,
  };
  it('render correctly with default props', () => {
    const { container } = render(<PasswordInput {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('render correctly with error message', async () => {
    const { getByText, getByRole } = render(
      <PasswordInput {...props} isInvalid errorMessage="Field is required" />,
    );
    const eyeButton = getByRole('button', {
      name: 'toggle password visibility',
    });

    await userEvent.click(eyeButton);

    expect(getByText('Field is required')).toBeInTheDocument();
  });

  it('should call onChange when PasswordInput value is changed', async () => {
    const { getByRole, getByPlaceholderText } = render(
      <PasswordInput {...props} />,
    );

    const eyeButton = getByRole('button', {
      name: 'toggle password visibility',
    });

    await userEvent.click(eyeButton);

    fireEvent.change(getByPlaceholderText('Email'), {
      target: { value: '1@Dzxcvb' },
    });

    await userEvent.click(eyeButton);

    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });
});
