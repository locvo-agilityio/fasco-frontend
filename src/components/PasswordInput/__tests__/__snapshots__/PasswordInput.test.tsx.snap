// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput component render correctly with default props 1`] = `
<div>
  <div
    class="flex w-full flex-wrap md:flex-nowrap mb-6 md:mb-0 gap-4"
  >
    <div
      class="group flex flex-col data-[hidden=true]:hidden w-full"
      data-filled="true"
      data-filled-within="true"
      data-has-elements="true"
      data-has-helper="true"
      data-has-label="true"
      data-slot="base"
    >
      <div
        class="w-full inline-flex tap-highlight-transparent px-3 !px-1 !pb-0 !gap-0 relative box-border border-b-medium shadow-[0_1px_0px_0_rgba(0,0,0,0.05)] border-default-200 !rounded-none hover:border-default-300 after:content-[''] after:w-0 after:origin-center after:absolute after:left-1/2 after:-translate-x-1/2 after:-bottom-[2px] after:h-[2px] group-data-[focus=true]:after:w-full min-h-10 rounded-medium flex-col items-start justify-center gap-0 transition-background motion-reduce:transition-none !duration-150 after:transition-width motion-reduce:after:transition-none h-14 py-2 after:bg-transparent"
        data-slot="input-wrapper"
        style="cursor: text;"
      >
        <label
          class="absolute z-10 pointer-events-none origin-top-left flex-shrink-0 rtl:origin-top-right subpixel-antialiased block cursor-text will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.small)/2_-_3.5px)] pe-2 max-w-full text-ellipsis overflow-hidden text-gray-500 text-base"
          data-slot="label"
          for="react-aria-:r0:"
          id="react-aria-:r1:"
        >
          Email
        </label>
        <div
          class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end pb-1.5"
          data-slot="inner-wrapper"
        >
          <input
            aria-label="Email"
            aria-labelledby="react-aria-:r0: react-aria-:r1:"
            class="w-full font-normal bg-transparent !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text group-data-[has-value=true]:text-gray-500 placeholder:text-gray-500 text-gray-500 text-base"
            data-has-end-content="true"
            data-slot="input"
            id="react-aria-:r0:"
            name="email"
            placeholder="Email"
            title=""
            type="password"
            value=""
          />
          <button
            aria-label="toggle password visibility"
            class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:!text-primary-foreground bg-transparent data-[hover=true]:!border-none text-secondary-400 border border-secondary-100 disabled:opacity-50 py-5 text-base rounded-base data-[pressed=true]:scale-[1] font-poppins w-fit focus:outline-none border-none data-[hover=true]:!bg-transparent"
            title="toggle password visibility"
            type="button"
          >
            <svg
              class="lucide lucide-eye-off pointer-events-none text-default-400"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10.733 5.076a10.744 10.744 0 0 1 11.205 6.575 1 1 0 0 1 0 .696 10.747 10.747 0 0 1-1.444 2.49"
              />
              <path
                d="M14.084 14.158a3 3 0 0 1-4.242-4.242"
              />
              <path
                d="M17.479 17.499a10.75 10.75 0 0 1-15.417-5.151 1 1 0 0 1 0-.696 10.75 10.75 0 0 1 4.446-5.143"
              />
              <path
                d="m2 2 20 20"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
